src/sources/kafka.rs

```diff
@@ -97,78 +96,95 @@ impl SourceConfig for KafkaSourceConfig {
+                                None => return Err(()), // skip messages with empty payload
                                 Some(Err(e)) => {
-                                    return Err(error!(message = "Cannot extract key", error = ?e))
+                                    return Err(
+                                        error!(message = "Cannot extract payload", error = ?e),
+                                    )
                                 }
-                                Some(Ok(key)) => {
-                                    log.insert(key_field.clone(), key);
+                                Some(Ok(payload)) => Bytes::from(payload),
+                            };
+                            let mut event = Event::new_empty_log();
+                            let log = event.as_mut_log();
+
+                            log.insert(event::log_schema().message_key().clone(), payload);
+
+                            // Extract timestamp from kafka message
+                            let timestamp = msg
+                                .timestamp()
+                                .to_millis()
+                                .and_then(|millis| Utc.timestamp_millis_opt(millis).latest())
+                                .unwrap_or_else(Utc::now);
+                            log.insert(event::log_schema().timestamp_key().clone(), timestamp);
+
+                            // Add source type
+                            log.insert(event::log_schema().source_type_key(), "kafka");
+
+                            if let Some(key_field) = &key_field {
+                                match msg.key_view::<[u8]>() {
+                                    None => (),
+                                    Some(Err(e)) => {
+                                        return Err(
+                                            error!(message = "Cannot extract key", error = ?e),
+                                        )
+                                    }
+                                    Some(Ok(key)) => {
+                                        log.insert(key_field.clone(), key);
+                                    }
                                 }
                             }
-                        }
 
-                        consumer_ref.store_offset(&msg).map_err(
-                            |e| error!(message = "Cannot store offset for the message", error = ?e),
-                        )?;
-                        Ok(event)
+                            consumer.store_offset(&msg).map_err(|e| error!(message = "Cannot store offset for the message", error = ?e))?;
+                            Ok(event)
+                        }
+                    }
+                }
+            })
+            // Not work =\
+            // .forward(
+            //     out.sink_compat()
+            //         .sink_map_err(|e| error!(message = "Error sending to sink", error = ?e)),
+            // )
```

Comment:
This looks like a useful note to future developers, but it's missing a "why". For example, "We can't use `.forward` here because …"

Reply:
Yeah, totally agree. Added such a comment.